librz/bin/format/java/class_attribute.c

```diff
@@ -5,6 +5,10 @@
 
 static ut8 *copy_buffer(RzBuffer *buf, st64 size, bool null_terminator) {
 	ut8 *buffer;
+	st64 buffer_size = rz_buf_size(buf);
```

Comment:
I was just reading this code for the first time, so might be a bit unrelated to this PR but i'm writing this anyway, since you wrote this code.

I think it is a bit risky to use `RzBuf` API in this way because it partially defeats its purpose, which is to never have to load the entire buffer in memory. This function can be used to do that though, because the `size` argument comes from the `attribute_length`, which is read as-is from the file, so an attacker can put any value there and force you to allocate ~4GB for each attribute.

Is it really necessary to have these buffer allocated? I see this is used in `ATTRIBUTE_TYPE_SOURCEDEBUGEXTENSION`, `ATTRIBUTE_TYPE_UNKNOWN`, `ATTRIBUTE_TYPE_CODE` but from a quick look i couldn't find where these buffers are used. You probably know better.

Another thing: what's the purpose of the `null_terminator` var? Can't we just always allocate it? Just asking.

Reply:
it really does allocate on file read? from the source code it doesn't seems to.

These extensions are used `ATTRIBUTE_TYPE_CODE` for the actual code, `ATTRIBUTE_TYPE_SOURCEDEBUGEXTENSION` for when you display all the details of the class.
`attribute_length` is always ut16, so whatever you place there is not that big on memory.

Reply:
`ut32 attribute_length;` and `attribute_length = rz_buf_read_be32(buf);` so I think it is actually 32bit, not 16.

> it really does allocate on file read? from the source code it doesn't seems to.

what do you mean? Here you are doing a `malloc(size)` where `size` is attacker-controlled so an attacker can essentially make this code allocate as much as he wants (up to almost 4GB).

> These extensions are used ATTRIBUTE_TYPE_CODE for the actual code, ATTRIBUTE_TYPE_SOURCEDEBUGEXTENSION for when you display all the details of the class.

Where exactly? I could not find them. As said, I'm not familiar enough with the code, so they might be used somewhere.

Reply:
oh, sorry i got confused on the 32bit, but anyway it's impossible to allocate 4GB if i check it against the file size. the max size for a class file (including everything) is 2GB.

Reply:
let's say that you even allocate 4GB (which is impossible with those changes), `if (!buffer || rz_buf_read(buf, buffer, size) < (size - 1)) {` will make sure to actually read 4GB. if it reads less it will drop the read and just say it failed to read the data

Reply:
about the `null_terminator` is because `ATTRIBUTE_TYPE_SOURCEDEBUGEXTENSION` is a string

Reply:
ok, i forgot to add `ATTRIBUTE_TYPE_SOURCEDEBUGEXTENSION` into the strings output. this is why you don't find a xref on that

Reply:
Shall we change:
```
if (!buffer || rz_buf_read(buf, buffer, size) < (size - 1)) {
```
to
```
if (!buffer || rz_buf_read(buf, buffer, size) != size) {
```
?


Anyway, my concerns as said were rather generic more than related to this specific PR. In general, it would be better if you avoid allocating all the memory in one shot but just keep using the RzBuf API.

Moreover, please double check that this buffer is really used and that `null_terminator` is really needed...

> about the null_terminator is because ATTRIBUTE_TYPE_SOURCEDEBUGEXTENSION is a string that doesn't end with a zero terminator

I see:
```
	attr->type = ATTRIBUTE_TYPE_SOURCEDEBUGEXTENSION;
	if (attr->attribute_length < 1) {
		return true;
	}
	st64 size = (st64)attr->attribute_length;
	attr->info = copy_buffer(buf, size, true);
```
so it seems like ATTRIBUTE_TYPE_SOURCEDEBUGEXTENSION wants the zero terminator.

Reply:
yes it is needed the null terminator, and no, i can't change that size, because it is always size-1 returned. dunno why..

Reply:
Just always add the null-terminator, I think there is no real need to put it only sometimes... Always having one byte more probably won't hurt anyway.

Reply:
ok. i'll patch this in another PR.